[tf-psa-crypto] psa_load_builtin_key_into_slot: prevent accessing the PSA storage if key ID is in volatile range

valeriosetti opened the pull request:
## Description

Resolves https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/488

## PR checklist

- [x] **changelog** provided.
- [x] **framework PR** not required
- [x] **mbedtls development PR** ~provided: https://github.com/Mbed-TLS/mbedtls/pull/10475~ not required
- [x] **mbedtls 3.6 PR** https://github.com/Mbed-TLS/mbedtls/pull/10473
- **tests**  added

valeriosetti committed:
core: replace psa_is_valid_key_id() with psa_key_id_is_user()

"psa_id_valid_key" seems a bit cryptic when it comes to its 2nd parameter, i.e.
"vendor". A new function is therefore proposed, "psa_key_id_is_user", which
together with "psa_key_id_is_builtin" (already existing) covers all use
cases of the removed function.

All usages of "psa_is_valid_key_id()" are then replaced by "psa_key_id_is_user"
or a combination of "psa_key_id_is_user" and "psa_key_id_is_builtin"
depending on the case.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
core: do not try to load persistent keys if ID is not in user range

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
include: psa: make psa_key_id_is_builtin() always available

psa_key_id_is_builtin() is only performing a simple check, so it won't cause
a big code increase to have it always enabled and not gated by
MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS.
OTOH it's handy to have it always enabled when working with PSA Core's
slot management operations.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on tests/suites/test_suite_psa_crypto_slot_management.data:
```diff
@@ -1,3 +1,21 @@
+Check key ID range: MBEDTLS_PSA_KEY_ID_BUILTIN_MIN
```
> If I've understood correctly the tests rely on these values being empty? Is there anything baked into the design currently and the future that guarantees this?

In the [test function](https://github.com/valeriosetti/TF-PSA-Crypto/blob/42919c46eb54849eea5f043039b0a0484a8eead2/tests/suites/test_suite_psa_crypto_slot_management.function#L137) that I added in this PR, I'm not allocating any key so almost all slots will be empty.
I'm saying "almost" because IIRC in some configurations where AES is accelerated then a volatile key is used by the PSA core for CTR-DRBG.

But this is not the point of this test. If you check where I placed hook functions [[1](https://github.com/valeriosetti/TF-PSA-Crypto/blob/42919c46eb54849eea5f043039b0a0484a8eead2/core/psa_crypto_slot_management.c#L713)] and [[2](https://github.com/valeriosetti/TF-PSA-Crypto/blob/42919c46eb54849eea5f043039b0a0484a8eead2/core/psa_crypto_slot_management.c#L674)] you'll see that these are really at the beginning of `psa_load_builtin_key_into_slot` and `psa_load_persistent_key_into_slot`, respectively. In fact I don't care of those function succeed; I only care if they go pass the initial check for the ID range or not.
Ideally I could also ignore the return value of `psa_get_and_lock_key_slot`, but in practice I need to know if by any chance the locking succeeded because then I had to unlock the key before exiting the function otherwise `PSA_DONE()` would have caused the test to fail due to slots still being locked at exit.

That's why I only took the top/bottom ID values for each range; I don't think that adding intermediate values would bring any benefit in this case.

Is this reasonable?

bjwtaylor replied:
So, I was wondering if in the future someone changes the design so the top or probably more likely bottom keys in one of the ranges is allocated would this break the test?

valeriosetti replied:
Ideally no, but now that you asked I realized that in case a volatile key slot is already existing `psa_load_builtin_key_into_slot` and `psa_load_persistent_key_into_slot` wouldn't been tested the way I thought in my test function.

bjwtaylor replied:
OK, probably not essential as we might never encounter it. However it's probably a relatively small change and might make it easier for someone in the future if they don't realise this tests is relying on that current feature of the design. What do you think?

valeriosetti replied:
Ok I updated the commit adding tests. Now it should be more precise and also clearer.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.data:
```diff
@@ -1,3 +1,21 @@
+
+Check key ID range: PSA_KEY_ID_VOLATILE_MAX
+check_key_id_value: PSA_KEY_ID_VOLATILE_MAX
+
```
Please add a test case with a key ID is not in the builtin range, not in the volatile range and not in the user range. (I think `PSA_KEY_ID_VENDOR_MAX` works, but the ranges are a bit complicated, we should make sure that whatever value we pick, the compilation or the test case fails if the value turns out to be in one of the known ranges after all.)

gilles-peskine-arm commented on core/psa_crypto_slot_management.c:
```diff
@@ -398,11 +378,7 @@ static psa_status_t psa_get_and_lock_key_slot_in_memory(
-            return PSA_ERROR_INVALID_HANDLE;
-        }
-
+    } else if (psa_key_id_is_user(key_id) || psa_key_id_is_builtin(key_id)) {
```
When a key ID is not in the builtin range, not in the volatile range and not in the user range, formerly, we would reject an attempt to create the key, but we would happily load the key from storage. This means that you could pre-provision a key by injecting it into the initial storage.

We never advertised this feature, so it's ok to remove it in a minor release. (Not in 3.6 LTS.) But I'd be more comfortable if there was a changelog entry. Something like

> Removals  
> The undocumented ability to load persistent keys outside the user key ID range has been removed. (This does not affect MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS.)

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -104,13 +104,78 @@ static size_t tiny_key_slice_length(size_t slice_idx)
+    mbedtls_test_hook_psa_load_builtin_key_into_slot = psa_load_builtin_key_into_slot_cb;
+    mbedtls_test_hook_psa_load_persistent_key_into_slot = psa_load_persistent_key_into_slot_cb;
+
+    psa_load_persistent_key_into_slot(&slot_persistent);
```
With dynamic key slots, parts of the key slot management code assumes that key slot pointers point inside the key slice indicated by `slot->slice_index`. I'm not sure if the particular functions called in this test are affected, but passing ad hoc slot structures is at least dodgy.

The CI seems happy about that, as far as I can tell: there's a memory leak but it doesn't seem related. (If there was a problem, I'd expect Asan to find an invalid pointer, not a memory leak.) But it still feels dodgy to me and I would really prefer the test to let the library code decide which slots it wants to access. Not necessarily through API functions, but perhaps starting with `
psa_get_and_lock_key_slot` (possibly with a hook in that function).

valeriosetti replied:
> I would really prefer the test to let the library code decide which slots it wants to access. Not necessarily through API functions, but perhaps starting with  psa_get_and_lock_key_slot (possibly with a hook in that function).

Initially I used `psa_get_and_lock_key_slot` indeed, but then while [discussing](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/492#discussion_r2436021863) with Ben I realized that doing so might have a corner case as follows. If a volatile key was really existing [^1] `psa_get_and_lock_key_slot_in_memory` would have succeeded thus preventing `psa_load_builtin_key_into_slot` and `psa_load_persistent_key_into_slot` from being called. The test case would be successful as well, but not the way I meant it to work: I want latter functions to be tested with all range of key IDs.
This is why I made `psa_load_builtin_key_into_slot` and `psa_load_persistent_key_into_slot` accessible from test code.

[^1]: Unless I'm mistaken this happens when AES is accelerated so that CTR-DRBG takes 1 slot during PSA Crypto initialization

valeriosetti replied:
I reset tests as they were previously in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/492/commits/baf3369f0b44b306aad5ec80c7cea39f5a603ee5. Is it better this way?

gilles-peskine-arm reviewed (changes requested):
Looks good to me for the library changes. There's a memory leak but I can't find any new code path that bypasses necessary cleanup. Maybe the new tests reveal a previously existing leak in an error case?

I don't understand the CI failures (but I haven't spent much time looking at them). The test code is dodgy but I don't see how it would cause a memory leak (I wouldn't be so surprised with an invalid pointer dereference or invalid `free()`).

valeriosetti committed:
changelog: persistent key loading support when ID is outside user ID range

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.data:
```diff
@@ -1,3 +1,24 @@
+Check key ID range: PSA_KEY_ID_VOLATILE_MAX
+check_key_id_value: PSA_KEY_ID_VOLATILE_MAX
+
+Check key ID range: PSA_KEY_ID_VENDOR_MAX (!volatile && !built-in && !user)
```
If we move some ranges around, `PSA_KEY_ID_VENDOR_MAX` could become part of the volatile range. I would prefer to have the test data assert which range the value fits in. That is:

```
void check_key_id_value(int key_id_arg, int is_builtin, int is_persistent)
```

and use `is_builtin` and `is_persistent` from the test data instead of calculating `builtin_attempts` and `persistent_attempts` in the test code.

valeriosetti replied:
I will modify test function and test cases as you suggested, but as for the `PSA_KEY_ID_VENDOR_MAX` I added that because I thought you asked for it [here](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/492#discussion_r2436993871). Did you misunderstood the request back then?

gilles-peskine-arm replied:
I did mean to suggest `PSA_KEY_ID_VENDOR_MAX`, yes (I went and fixed the name, but you got what I meant). But I wanted to have some check that the value is indeed neither in the builtin range nor in the persistent range. In fact, for a long time, we had volatile keys at the top of the vendor range, rather than the bottom (`#define PSA_KEY_ID_VOLATILE_MAX  PSA_KEY_ID_VENDOR_MAX`). We changed that when we added the dynamic key store with a larger range, because there was no room to grow because built-in keys came just before, now it's at the bottom. In my tests, I check that the chosen value for this test case is indeed in none of the meaningful ranges through assertions on `KEY_ID_OUTSIDE_DEFINED_RANGES`.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_slot_management.function:
```diff
@@ -104,13 +104,76 @@ static size_t tiny_key_slice_length(size_t slice_idx)
+    mbedtls_test_hook_psa_load_builtin_key_into_slot = psa_load_builtin_key_into_slot_cb;
+    mbedtls_test_hook_psa_load_persistent_key_into_slot = psa_load_persistent_key_into_slot_cb;
+
+    status = psa_get_and_lock_key_slot(key_id, &slot);
```
I'm still [bothered by fiddling with slots](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/492#discussion_r2437037763). It seems to work (in the sense that Asan doesn't find anything wrong), but I don't have a good sense of what you may and may not with dynamic key slots (and I wrote that code!). I'm need to think about it a little more.

gilles-peskine-arm replied:
I wrote some tests that check what happens if the application tries to access a key and the storage contains something bad (either an unexpected file for a key ID outside the range for persistent keys, or some bad file content): "Test with bad data in storage: key IDs"  2828df12289958d7ec167a81d1fe2ea8261fa861 on https://github.com/gilles-peskine-arm/TF-PSA-Crypto/tree/psa-test-load-key-id-ranges-1.0 . [With your fix](https://github.com/gilles-peskine-arm/TF-PSA-Crypto/tree/psa-test-load-key-id-ranges-builtin-fix), these tests pass. (The tests with bad file content that I added in another commit fail, but that's a separate matter.)

What do you think about these tests in terms of what they cover? On my side, they give me enough confidence that I think we don't need to fiddle with slots.

gilles-peskine-arm reviewed (commented):
As of aa6b013b92f39df9e439f54a64243f2f3a6cc61d, this looks good for the library code, but I'm still uncomfortable about the testing:

* Because it's fiddling with slots and that feels risky.
* Because it's asserting on functions that are deep inside the implementation, and I would prefer to assert in places with a clearer contract, ideally on public functions (in this system state, what does `psa_get_key_attributes` report? Possibly asserting on whether it calls an ITS function.).

I'm going to think about this at least overnight.

gilles-peskine-arm committed:
Test key identifier ranges

Add some assertions on the various ranges of key identifiers to ensure that
they're disjoint and they comply with documented guarantees.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Test with bad data in storage: key IDs

Test what happens when the application tries to access a key and the storage
contains something invalid: either a corrupted file, or a key ID that's
outside the standard range for persistent keys.

Coverage of APIs in this commit:

* `psa_get_key_attributes()` (generally as a proxy for any key access);
* `psa_export_key()` (minor, but does provide some coverage of what happens
  if only the key material is corrupted);
* `psa_destroy_key()`, which hopefully should work even for a corrupted file.

Coverage of key IDs in this commit:

* Key IDs in various ranges: user (i.e. the normal range for persistent
  keys), builtin, volatile, reserved file ID, none of the above.
* Includes coverage for nonzero owner ID.

No coverage of corrupted files in this commit.

Assert the behavior that I think is the right thing. Subsequent commits will
reconcile the library behavior with the code as needed.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

valeriosetti commented:
I partially rewrote the history of this PR dropping my commits adding previous testing and replacing them with Gilles' proposal.

gilles-peskine-arm reviewed (changes requested):
Looks good to me at b55a15aeca6566cd420e8ee529fe9dea66c4f055, which add non-regression tests for the bug that is fixed here. The next commit "Test with bad data in storage: bad key metadata" reveals a separate bug https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/544 which is out of scope here. Please remove that commit.

gilles-peskine-arm commented:
I think we should backport the tests, and part of the bug fix. Attempting to access a key ID in the volatile range should definitely not try to access storage: that would lead to behavior that's very hard to predict, depending on which volatile key IDs are in use. Other parts of the vendor range are more ambiguous: we never officially said integrators could use them, but if they did use them and it works for them, we shouldn't break that in an LTS. So I think in 3.6 we should keep allowing storage access everywhere in the vendor range except the volatile range.

gilles-peskine-arm reviewed (dismissed):
LGTM

bjwtaylor reviewed (dismissed):
There seems to be code style issues, but otherwise looks good.

valeriosetti committed:
tests: suite_psa_crypto_persistent: fix formatting

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
> There seems to be code style issues, but otherwise looks good.

I didn't notice that, thanks for the heads up! It should be fixed now.

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.function:
```diff
@@ -452,6 +452,17 @@ void load_primed_storage(int32_t owner_id,
+    /* Skip the test if we happened to pick a builtin key used by the driver. */
+    if (psa_key_id_is_builtin(key_id)) {
+        if (psa_get_key_attributes(key_id, &attributes) == PSA_SUCCESS) {
+            psa_destroy_key(key_id);
```
I don't understand: if the key is already in use, why would we destroy it?

valeriosetti replied:
Initially `psa_destroy_key` wasn't there indeed and I had to add it because `PSA_DONE` was complaining that there was a key slot unclosed at the end. After some debug it turned out that `psa_get_key_attributes` actually loads a slot and it terminates with "no reader" for that slot, but the slot not being wiped out (just in "pending deletion" state). So I added the call to `psa_destroy_key` to clean it up. Was perhaps a call to `psa_wipe_key_slot` better?

valeriosetti replied:
The key is not really in use by anyone as far as I can tell. It's "known to the driver" as builtin key and it gets loaded into the slot when `get_attributes` is called.

gilles-peskine-arm replied:
Oh, I see. This makes sense but I'm going to dig a little into what we're doing with built-in keys. At the very least, please explain what's going on in a comment.

gilles-peskine-arm replied:
Thinking some more about this: the goal of the test is to validate what happens when a key file exists in storage, depending on what else might apply to the key ID. The original bug was about the case where no key by this ID should exist, in particular when the key ID is in the volatile range or in the built-in range. But for good test coverage, we should also handle the case when a file exists in storage _and_ a built-in key exists. We would need different assertions for this case, since the expected behavior is different.

So I think that when the key ID is in the built-in range, the test should allow the key to exist, but it should check that the key is not what we put in storage. I'm not sure exactly how to do this. We can call `mbedtls_psa_platform_get_builtin_key` to check if the key ID is a built-in key provided by our test harness, with the downside that the test won't work on a real platform. We can also just allow `psa_get_key_attributes` to return 0 instead of `INVALID_HANDLE` for a built-in key, and in that case ensure that the key has different attributes or different content compared to the test case data.

From the perspective of test coverage, there are three cases:

* Built-in keys are disabled.
* Built-in keys are enabled, but there is no built-in key with this ID.
* Built-in keys are enabled, and there is a built-in key with this ID.

Ideally these should be different test cases, so that we can validate that all are covered. But that might be awkward to arrange, at the very least it would mean that some variation of https://github.com/Mbed-TLS/mbedtls/pull/10475 remains necessary even though we wouldn't be skipping some testing.

gilles-peskine-arm replied:
On third thoughts, the testing here focuses on what happens if we try to access a key that shouldn't exist. For example, there's no test case with an existing volatile key, and an existing file for the same key ID. Arguably, there should be one. But also arguably, it's out of scope: the original problem is to validate that if a key shouldn't exist then the library doesn't try to load it from storage when it shouldn't. It's still a test gap: we have test cases that validate that we can access volatile and built-in keys, but they don't put a file with the corresponding ID in storage. So we're assuming that if attempting to access a non-existing built-in or volatile key ID does not cause it to be loaded from storage, then this is still true for an existing key. That's plausible, but still not ideal.

So I would prefer to extend test coverage to have a case with an existing volatile key and a case with an existing built-in key. But it's tricky to arrange and borderline in terms of scope, so I'd be ok with the current code, with more explanation in a comment.

gilles-peskine-arm replied:
What do you think of this? https://github.com/gilles-peskine-arm/TF-PSA-Crypto/tree/psa-test-load-key-id-ranges-test-existing

When the volatile/builtin key exists, instead of skipping, I change the expectations to be that we're accessing the existing key and not the file.

It does make the code of the test function even more complicated, but I like the resulting test coverage.

I'd like it even more if “existing built-in key” and “non-existing built-in key” were separate test cases, and likewise volatile, but I can't think of a way to arrange that that isn't very fiddly.

valeriosetti commented:
I created https://github.com/Mbed-TLS/mbedtls/pull/10475 to resolve disparities in driver testing.

gilles-peskine-arm committed:
Test with bad data in storage: support existing built-in or volatile key

When testing what happens with when accessing a key ID in the built-in or
volatile range and a file exists in storage, we were skipping the test case
when the key existed. When the volatile or built-in key exists, the
expectations on the test case are wrong, but the test case is still useful:
we should ensure that the existence of the file doesn't somehow prevent
access to the built-in or volatile key. So, instead of skipping, change the
test assertions on the fly to ensure that we are accessing the existing key.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.function:
```diff
@@ -338,3 +406,169 @@ exit:
+               expected_export_status);
+    if (expected_export_status == PSA_SUCCESS) {
+        if (key_already_existed) {
+            /* The key already existed. We don't know what it is,
+             * but check that it is not what we put in storage. */
+            TEST_ASSERT(key_data_length != payload_length ||
+                        memcmp(key_data, payload, payload_length));
```
Note that this would be a false negative if a built-in key happens to exist with the same key material (length and content) as what we put in the test data. Currently, the payload in the relevant test data (test cases with a key ID that's in the range of built-in keys) is `"4c"` because that was an existing key representation that I just copied. For more robustness, we could replace it with something more unique like `"6b6579206d6174657269616c20696e2070657273697374656e742073746f72616765"` (and adjust the key bit-size accordingly).

gilles-peskine-arm reviewed (approved):
Looks good to me. (There's code that I wrote here, which I Valerio has accepted to use, so this conveys our joint approval.)

(As discussed in Slack, there's something in the test that I meant to do before pushing and forgot, but it isn't critical.)

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto_persistent_key.function:
```diff
@@ -338,3 +406,169 @@ exit:
+exit:
+    psa_reset_key_attributes(&attributes);
+    PSA_DONE();
+    psa_its_remove(uid);
```
I forgot that [this removal, too, must not be done for reserved files](https://github.com/Mbed-TLS/mbedtls/pull/10473#discussion_r2485803296).


```suggestion
    if (!KEY_ID_IN_RESERVED_FILE_ID_RANGE(key_id_arg) {
        /* The key ID corresponds to a reserved file (e.g. transaction
         * log or entropy seed). Don't corrupt that file. */
        psa_its_remove(uid);
    }
```

gilles-peskine-arm reviewed (changes requested):
We don't currently implement any reserved file IDs, but when we do in the future, this test's cleanup code would mess them up. I forgot that in 5dc4aca9d5c067556f4932de84487f2bc00a5288.

valeriosetti committed:
tests: suite_psa_crypto_persistent_key: fix load_primed_storage

Do not remove keys that belong to the reserved range.

Signed-off-by: Valerio Setti <valerio.setti@nordicsemi.no>

gilles-peskine-arm reviewed (approved):
LGTM

mpg reviewed (approved):
LGTM, (Ben had approved up to 5dc4aca9d5c067556f4932de84487f2bc00a5288 so I've only reviewed the final commit myself.)
